File: solutions/python/min-heap/04_extract/min_heap.py
class MinHeap:
    def __init__(self) -> None:
        self._items = []

    def insert(self, value) -> None:
        self._items.append(value)
        self._bubble_up(len(self._items) - 1)

    def peek_min(self):
        if not self._items:
            return None
        return self._items[0]

    def extract_min(self):
        if not self._items:
            return None
        minimum = self._items[0]
        end = self._items.pop()
        if self._items:
            self._items[0] = end
            # Bubble down to be implemented next.
        return minimum

    def _bubble_up(self, index: int) -> None:
        current = index
        while current > 0:
            parent = (current - 1) // 2
            if self._items[parent] <= self._items[current]:
                break
            self._items[parent], self._items[current] = (
                self._items[current],
                self._items[parent],
            )
            current = parent

File: solutions/javascript/min-heap/04-extract/minHeap.js
class MinHeap {
  constructor() {
    this._items = [];
  }

  insert(value) {
    this._items.push(value);
    this._bubbleUp(this._items.length - 1);
  }

  peekMin() {
    if (this._items.length === 0) {
      return undefined;
    }
    return this._items[0];
  }

  extractMin() {
    if (this._items.length === 0) {
      return undefined;
    }
    const min = this._items[0];
    const end = this._items.pop();
    if (this._items.length > 0) {
      this._items[0] = end;
      // Bubble down not implemented yet.
    }
    return min;
  }

  _bubbleUp(index) {
    let current = index;
    while (current > 0) {
      const parent = Math.floor((current - 1) / 2);
      if (this._items[parent] <= this._items[current]) {
        break;
      }
      const temp = this._items[parent];
      this._items[parent] = this._items[current];
      this._items[current] = temp;
      current = parent;
    }
  }
}

module.exports = MinHeap;
